### ISSUE TITLE
Make both rubocop and the NilCheck detector happy

### ISSUE THREAD

**Issue by troessner** [troessner/reek#1375]
Coming from #1374:

With making the 2.3. the minimal Ruby version we had to add the following exception to rubocop:

```Yaml
Style/SafeNavigation:
  Exclude:
    - lib/reek/ast/node.rb
    - lib/reek/ast/sexp_extensions/module.rb
```

Without this exception rubocop will complain:

```
Offenses:

lib/reek/ast/node.rb:31:9: C: [Corrected] Style/SafeNavigation: Use safe navigation (&.) instead of checking if an object exists before calling the method.
        loc && loc.line
        ^^^^^^^^^^^^^^^
lib/reek/ast/sexp_extensions/module.rb:105:11: C: [Corrected] Style/SafeNavigation: Use safe navigation (&.) instead of checking if an object exists before calling the method.
          call && call.module_creation_call?
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Letting Rubocop auto-fix is will cause this line:

```Ruby
loc && loc.line
```

to become this line:

```Ruby
loc&.line
```

The problem is that now our reek quality check fails with:

```
Failures:

  1) Reek source code lib/reek/ast/node.rb has no smells
     Failure/Error: expect(pathname).not_to reek

       Expected no smells, but got:
         NilCheck: Reek::AST::Node#line performs a nil-check
     # ./spec/quality/reek_source_spec.rb:7:in `block (4 levels) in <top (required)>'

  2) Reek source code lib/reek/ast/sexp_extensions/module.rb has no smells
     Failure/Error: expect(pathname).not_to reek

       Expected no smells, but got:
         NilCheck: Reek::AST::SexpExtensions::CasgnNode#defines_module? performs a nil-check
     # ./spec/quality/reek_source_spec.rb:7:in `block (4 levels) in <top (required)>'

Finished in 2.42 seconds (files took 0.87515 seconds to load)
128 examples, 2 failures

Failed examples:

rspec ./spec/quality/reek_source_spec.rb[1:77:1] # Reek source code lib/reek/ast/node.rb has no smells
rspec ./spec/quality/reek_source_spec.rb[1:84:1] # Reek source code lib/reek/ast/sexp_extensions/module.rb has no smells
```

I'd rather not defer our Reek 5 release any longer because of that but rather fix this in a follow-up issue afterwards.

**Comment by Mifrill** [comment]
@troessner any updates on this?

**Comment by troessner** [comment]
I'm not planning to work on this any time, maybe @mvz is in the mood ;)

**Comment by Mifrill** [comment]
@troessner okay, thanks for reply. Just want ask you opinion about it: do you still think it should fixed or it's really the code-smells indicator about some of problem deeper in code?

**Comment by troessner** [comment]
No, I think this is something we should fix in Reek.

**Comment by mvz** [comment]
Technically, Reek's judgment is correct since `foo&.bar` is not equivalent to `foo && foo.bar`, but to `foo.bar unless foo.nil?`. This means RuboCop's corrected version raises an exception if `foo` is `false`.

The question is, does this difference matter in practice? It seems not since RuboCop sweeps it under the rug.

TL;DR: The simplest solution is to not consider `foo&.bar` a case of NilCheck.

**Comment by troessner** [comment]
@mvz are you going to release those changes?

**Comment by mvz** [comment]
I'm holding off a little to see if there are any more breaking changes for the 6.0 release. We should probably pick a deadline for that so it won't take ages.
